engine/servermaster/jobop/operator.go

```diff
@@ -0,0 +1,122 @@
+type JobOperator interface {
+	MarkJobCanceling(ctx context.Context, jobID string) error
+	MarkJobCanceled(ctx context.Context, jobID string) error
+	Tick(ctx context.Context) error
+}
+
+// JobOperatorImpl implements JobOperator
+type JobOperatorImpl struct {
+	frameMetaClient pkgOrm.Client
+	router          operateRouter
+}
+
+// NewJobOperatorImpl creates a new JobOperatorImpl
+func NewJobOperatorImpl(cli pkgOrm.Client, router operateRouter) *JobOperatorImpl {
+	return &JobOperatorImpl{
+		frameMetaClient: cli,
+		router:          router,
+	}
+}
+
+func (oper *JobOperatorImpl) updateJobOperationStatus(
+	ctx context.Context, jobID string, op ormModel.JobOpStatus,
+) error {
+	var ormFn func(ctx context.Context, JobID string) (pkgOrm.Result, error)
+	switch op {
+	case ormModel.JobOpStatusNoop:
+		ormFn = oper.frameMetaClient.SetJobNoop
+	case ormModel.JobOpStatusCanceling:
+		ormFn = oper.frameMetaClient.SetJobCanceling
+	case ormModel.JobOpStatusCanceled:
+		ormFn = oper.frameMetaClient.SetJobCanceled
+	default:
+		log.Panic("unexpected job operate", zap.Any("op", op))
+	}
+	if result, err := ormFn(ctx, jobID); err != nil {
+		return err
+	} else if result.RowsAffected() == 0 {
+		log.Info("job status is already set", zap.String("job-id", jobID), zap.Any("op", op))
+	}
+	return nil
+}
+
+// MarkJobNoop implements JobOperator.MarkJobNoop
+func (oper *JobOperatorImpl) MarkJobNoop(ctx context.Context, jobID string) error {
+	return oper.updateJobOperationStatus(ctx, jobID, ormModel.JobOpStatusNoop)
+}
+
+// MarkJobCanceling implements JobOperator.MarkJobCanceling
+func (oper *JobOperatorImpl) MarkJobCanceling(ctx context.Context, jobID string) error {
+	return oper.updateJobOperationStatus(ctx, jobID, ormModel.JobOpStatusCanceling)
+}
+
+// MarkJobCanceled implements JobOperator.MarkJobCanceled
+func (oper *JobOperatorImpl) MarkJobCanceled(ctx context.Context, jobID string) error {
+	return oper.updateJobOperationStatus(ctx, jobID, ormModel.JobOpStatusCanceled)
+}
+
+// Tick implements JobOperator.Tick
+func (oper *JobOperatorImpl) Tick(ctx context.Context) error {
+	ops, err := oper.frameMetaClient.QueryJobOpsByStatus(ctx, ormModel.JobOpStatusCanceling)
+	if err != nil {
+		return err
+	}
+	for _, op := range ops {
+		err := oper.router.SendCancelJobMessage(ctx, op.JobID)
+		if err != nil {
+			log.Warn("send cancel message to job master failed",
+				zap.String("job-id", op.JobID), zap.Error(err))
+			if err := oper.checkJobStatus(ctx, op.JobID); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (oper *JobOperatorImpl) checkJobStatus(ctx context.Context, jobID string) error {
+	meta, err := oper.frameMetaClient.GetJobByID(ctx, jobID)
+	if err != nil {
+		if pkgOrm.IsNotFoundError(err) {
+			log.Warn("found orphan job operation", zap.String("job-id", jobID))
+			return oper.MarkJobNoop(ctx, jobID)
+		}
+		return err
+	}
+	switch meta.StatusCode {
+	case frameworkModel.MasterStatusFinished, frameworkModel.MasterStatusStopped:
```

Comment:
If a job is marked `canceling` and fail at last, I think we can also mark the op `Canceled` since `failed` is a terminated state.

Reply:
Add a comment, will be added in another PR.